gcc/rust/resolve/rust-ast-resolve-expr.cc

```diff
@@ -384,7 +389,18 @@ ResolveExpr::visit (AST::BreakExpr &expr)
     }
 
   if (expr.has_break_expr ())
-    ResolveExpr::go (expr.get_break_expr ().get (), prefix, canonical_prefix);
+    {
+      bool funny_error = false;
+      AST::Expr &break_expr = *expr.get_break_expr ().get ();
+      if (break_expr.get_ast_kind () == AST::Kind::IDENTIFIER)
+	{
+	  std::string ident
+	    = static_cast<AST::IdentifierExpr &> (break_expr).as_string ();
+	  if (ident == "rust" || ident == "gcc")
+	    funny_error = true;
```

Comment:
Why not just emit the error here without the need for the funny_error flag?

Reply:
This was [suggested by @CohenArthur](https://github.com/Rust-GCC/gccrs/issues/1996#issuecomment-1491915749). Basically, at this point we don't yet know that there will be an error; we only want to emit an error if the identifier later fails to resolve. For example (this is `break-rust1.rs`, one of the test case added by this PR):

```rust
let rust = "crab";
let res = loop {
    break rust;
};
```

This should still compile just fine. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8edc59d0dcbcd30231eb182aae97a6ca). (Actually, as you can see by commenting out the definition of `rust` in the playground, rustc still emits the "boring" error if it encounters `break rust` inside a loop, because they do it at a different stage (when recovering from `break` outside of a loop). I didn't replicate that; gccrs will still ICE on that example.)

Reply:
Hopefully the comment answers this now :)